### PR TITLE
executor: enhance memory tracking for IndexLookup

### LINKED ISSUE #45901: enhance the memory track for IndexLookup
## Enhancement
截屏2023-08-08 15 25 41

The following elements should be tracked:
https://github.com/pingcap/tidb/blob/3296b5e6032b1c54f496ffdda8fb13f9be46b8b5/executor/distsql.go#L1139-L1146

https://github.com/pingcap/tidb/blob/3296b5e6032b1c54f496ffdda8fb13f9be46b8b5/executor/distsql.go#L1241-L1245

https://github.com/pingcap/tidb/blob/3296b5e6032b1c54f496ffdda8fb13f9be46b8b5/executor/builder.go#L4629

### PR DESCRIPTION
**Pull request by solotzg**
<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
<!--

Please create an issue first to describe the problem.

There MUST be one line starting with "Issue Number:  " and
linking the relevant issues via the "close" or "ref".

For more info, check https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/contribute-code.html#referring-to-an-issue.

-->

Issue Number: close #45901

Problem Summary:

### What changed and how does it work?

- track more memory info in `func (w *tableWorker) executeTask` and `func (e *IndexMergeReaderExecutor) buildFinalTableReader`
  - track `handles`, `indexOrder`, `duplicatedIndexOrder` of `lookupTableTask`
  - track `TableReaderExecutor` 

### Tests

env
```
sysbench --db-driver=mysql --table-size=10000000 oltp_read_write prepare 

mysql> select count(1) from sbtest1;
+----------+
| count(1) |
+----------+
| 10000000 |
+----------+
1 row in set (1.07 sec)
```

sql

```
use loadtest;
explain analyze
SELECT MIN(k) AS start_key,
    MAX(k) AS end_key,
    COUNT(*) AS page_size
FROM (
        SELECT a.*,
            rank () over (
                ORDER BY k
            ) AS row_num
        FROM sbtest1 a
        WHERE c <> ''
    ) t;
```

test result:

original: 

- mem usage about `IndexLookUp` is about `51.6 MB`

```
mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                      |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.6.0-alpha-443-g3379450cec
Edition: Community
Git Commit Hash: 3379450cec573c3ab70d9601667072ea45e18ea0
Git Branch: 45901
UTC Build Time: 2023-12-13 08:46:04
GoVersion: go1.21.0
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

+-------------------------------------+-------------+----------+-----------+-----------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------+----------+------+
| id                                  | estRows     | actRows  | task      | access object         | execution info                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | operator info                                                                                              | memory   | disk |
+-------------------------------------+-------------+----------+-----------+-----------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------+----------+------+
| HashAgg_10                          | 1.00        | 1        | root      |                       | time:8.09s, loops:2, RU:117291.687041, partial_worker:{wall_time:8.091658309s, concurrency:5, task_num:9766, tot_wait:39.637924386s, tot_exec:818.402781ms, tot_time:40.458064551s, max:8.091633581s, p95:8.091633581s}, final_worker:{wall_time:8.09166921s, concurrency:5, task_num:20, tot_wait:40.458235164s, tot_exec:15.717µs, tot_time:40.458258557s, max:8.091656428s, p95:8.091656428s}                                                                                                                                                                                                                                                             | funcs:min(loadtest.sbtest1.k)->Column#7, funcs:max(loadtest.sbtest1.k)->Column#8, funcs:count(1)->Column#9 | 1.72 MB  | N/A  |
| └─Window_12                         | 9999998.99  | 10000000 | root      |                       | time:7.93s, loops:9767                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | rank()->Column#6 over(order by loadtest.sbtest1.k)                                                         | N/A      | N/A  |
|   └─Projection_17                   | 9999998.99  | 10000000 | root      |                       | time:6.66s, loops:9767, Concurrency:5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | loadtest.sbtest1.k, loadtest.sbtest1.c                                                                     | 801.1 KB | N/A  |
|     └─IndexLookUp_16                | 9999998.99  | 10000000 | root      |                       | time:6.82s, loops:9767, index_task: {total_time: 7.2s, fetch_handle: 718.7ms, build: 1.54s, wait: 4.94s}, table_task: {total_time: 35.5s, num: 493, concurrency: 5}, next: {wait_index: 3.24ms, wait_table_lookup_build: 589µs, wait_table_lookup_resp: 2.61s}                                                                                                                                                                                                                                                                                                                                                                                               |                                                                                                            | 51.6 MB  | N/A  |
|       ├─IndexFullScan_13(Build)     | 10000000.00 | 10000000 | cop[tikv] | table:a, index:k_1(k) | time:146.7ms, loops:9794, cop_task: {num: 287, max: 59ms, min: 514.5µs, avg: 20.2ms, p95: 36.9ms, max_proc_keys: 50144, p95_proc_keys: 50144, tot_proc: 5.2s, tot_wait: 14.6ms, rpc_num: 287, rpc_time: 5.78s, copr_cache: disabled, build_task_duration: 57.4µs, max_distsql_concurrency: 2}, tikv_task:{proc max:41ms, min:0s, avg: 18ms, p80:28ms, p95:34ms, iters:10909, tasks:287}, scan_detail: {total_process_keys: 10000000, total_process_keys_size: 460000000, total_keys: 10000287, get_snapshot_time: 4.14ms, rocksdb: {key_skipped_count: 10000000, block: {cache_hit_count: 15135}}}                                                           | keep order:true                                                                                            | N/A      | N/A  |
|       └─Selection_15(Probe)         | 9999998.99  | 10000000 | cop[tikv] |                       | time:26.1s, loops:10273, cop_task: {num: 12805, max: 72.5ms, min: 0s, avg: 22.2ms, p95: 32.4ms, max_proc_keys: 948, p95_proc_keys: 885, tot_proc: 3m33.3s, tot_wait: 722.6ms, rpc_num: 12791, rpc_time: 4m44.2s, copr_cache: disabled, build_task_duration: 405.4ms, max_distsql_concurrency: 15, max_extra_concurrency: 1, store_batch_num: 14}, tikv_task:{proc max:0s, min:0s, avg: 17.7ms, p80:22ms, p95:28ms, iters:62692, tasks:12805}, scan_detail: {total_process_keys: 10000000, total_process_keys_size: 2240000000, total_keys: 20020457, get_snapshot_time: 143.3ms, rocksdb: {key_skipped_count: 10040952, block: {cache_hit_count: 36808923}}} | ne(loadtest.sbtest1.c, "")                                                                                 | N/A      | N/A  |
|         └─TableRowIDScan_14         | 10000000.00 | 10000000 | cop[tikv] | table:a               | tikv_task:{proc max:0s, min:0s, avg: 17.5ms, p80:22ms, p95:28ms, iters:62692, tasks:12805}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | keep order:false                                                                                           | N/A      | N/A  |
+-------------------------------------+-------------+----------+-----------+-----------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------+----------+------+
7 rows in set (8.09 sec)
```

PR#49336

- mem usage about `IndexLookUp` is about `88.2 MB`, which is closer to the real value.

```
+-------------------------------------+-------------+----------+-----------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------+----------+------+
| id                                  | estRows     | actRows  | task      | access object         | execution info                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | operator info                                                                                              | memory   | disk |
+-------------------------------------+-------------+----------+-----------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------+----------+------+
| HashAgg_10                          | 1.00        | 1        | root      |                       | time:8.26s, loops:2, RU:122572.111836, partial_worker:{wall_time:8.257561221s, concurrency:5, task_num:9766, tot_wait:40.290672285s, tot_exec:994.592439ms, tot_time:41.287112739s, max:8.257456854s, p95:8.257456854s}, final_worker:{wall_time:8.257581094s, concurrency:5, task_num:20, tot_wait:41.287370609s, tot_exec:15.519µs, tot_time:41.287398669s, max:8.257485223s, p95:8.257485223s}                                                                                                                                                                                                                                                         | funcs:min(loadtest.sbtest1.k)->Column#7, funcs:max(loadtest.sbtest1.k)->Column#8, funcs:count(1)->Column#9 | 1.72 MB  | N/A  |
| └─Window_12                         | 9999998.99  | 10000000 | root      |                       | time:8.06s, loops:9767                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | rank()->Column#6 over(order by loadtest.sbtest1.k)                                                         | N/A      | N/A  |
|   └─Projection_17                   | 9999998.99  | 10000000 | root      |                       | time:6.85s, loops:9767, Concurrency:5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | loadtest.sbtest1.k, loadtest.sbtest1.c                                                                     | 801.1 KB | N/A  |
|     └─IndexLookUp_16                | 9999998.99  | 10000000 | root      |                       | time:6.96s, loops:9767, index_task: {total_time: 7.44s, fetch_handle: 663.7ms, build: 1.35s, wait: 5.42s}, table_task: {total_time: 36.7s, num: 493, concurrency: 5}, next: {wait_index: 12.8ms, wait_table_lookup_build: 5.61ms, wait_table_lookup_resp: 3.12s}                                                                                                                                                                                                                                                                                                                                                                                          |                                                                                                            | 88.2 MB  | N/A  |
|       ├─IndexFullScan_13(Build)     | 10000000.00 | 10000000 | cop[tikv] | table:a, index:k_1(k) | time:134.2ms, loops:9794, cop_task: {num: 287, max: 56.1ms, min: 605.2µs, avg: 21.9ms, p95: 39.7ms, max_proc_keys: 50144, p95_proc_keys: 50144, tot_proc: 5.57s, tot_wait: 17.3ms, rpc_num: 287, rpc_time: 6.27s, copr_cache: disabled, build_task_duration: 75.4µs, max_distsql_concurrency: 2}, tikv_task:{proc max:51ms, min:0s, avg: 19.2ms, p80:30ms, p95:35ms, iters:10909, tasks:287}, scan_detail: {total_process_keys: 10000000, total_process_keys_size: 460000000, total_keys: 10000287, get_snapshot_time: 5.6ms, rocksdb: {key_skipped_count: 10000000, block: {cache_hit_count: 15135}}}                                                    | keep order:true                                                                                            | N/A      | N/A  |
|       └─Selection_15(Probe)         | 9999998.99  | 10000000 | cop[tikv] |                       | time:27.7s, loops:10273, cop_task: {num: 12805, max: 68.6ms, min: 0s, avg: 23.9ms, p95: 37.1ms, max_proc_keys: 948, p95_proc_keys: 885, tot_proc: 3m48.7s, tot_wait: 788.1ms, rpc_num: 12791, rpc_time: 5m6.4s, copr_cache: disabled, build_task_duration: 393.4ms, max_distsql_concurrency: 15, max_extra_concurrency: 1, store_batch_num: 14}, tikv_task:{proc max:0s, min:0s, avg: 19ms, p80:24ms, p95:31ms, iters:62692, tasks:12805}, scan_detail: {total_process_keys: 10000000, total_process_keys_size: 2240000000, total_keys: 20020457, get_snapshot_time: 166.7ms, rocksdb: {key_skipped_count: 10040952, block: {cache_hit_count: 36808923}}} | ne(loadtest.sbtest1.c, "")                                                                                 | N/A      | N/A  |
|         └─TableRowIDScan_14         | 10000000.00 | 10000000 | cop[tikv] | table:a               | tikv_task:{proc max:0s, min:0s, avg: 18.8ms, p80:24ms, p95:31ms, iters:62692, tasks:12805}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | keep order:false                                                                                           | N/A      | N/A  |
+-------------------------------------+-------------+----------+-----------+-----------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------+----------+------+
7 rows in set (8.27 sec)
```


### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > <!-- Or your custom  "No need to test" reasons -->

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

<!-- compatibility change, improvement, bugfix, and new feature need a release note -->

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by solotzg** [comment]
cc @XuHuaiyu @wshwsh12

File: pkg/kv/key.go [hunk 1]
```diff
@@ -404,6 +406,9 @@ type strHandleVal struct {
 	val interface{}
 }
 
+const sizeofHandleMap = int64(unsafe.Sizeof(*(*HandleMap)(nil)))
```
Comment (XuHuaiyu) [inline pkg/kv/key.go:409]: We need to track the memory usage of the underlying structure of golang map

refer: https://github.com/pingcap/tidb/blob/master/pkg/kv/key.go#L527

**Review by XuHuaiyu — COMMENTED** [review]
~ I do not find where the memory usage of `codec.EncodeKey` and `tablecodec.EncodeRowKey` is tracked. ~
image

File: pkg/executor/table_reader.go [hunk 2]
```diff
@@ -138,6 +142,21 @@ func (e *TableReaderExecutor) setDummy() {
 	e.dummy = true
 }
 
+func (e *TableReaderExecutor) memUsage() int64 {
```
Comment (wshwsh12) [inline pkg/executor/table_reader.go:145]: We should assign a tracker to TableReaderExecutor and attach it to IndexLookup, and TableReader tracks the memory usage itself.

Comment (solotzg) [reply pkg/executor/table_reader.go:145]: https://github.com/pingcap/tidb/blob/c1cae24a87792055e96a213afc9f1459a183b767/pkg/executor/distsql.go#L769-L783

The memTracker of `tableWorker` has been attached to memTracker of `IndexLookUpExecutor `

**Review by XuHuaiyu — APPROVED** [review]

**Review by wshwsh12 — APPROVED** [review]

**Pull request merged**